tests/trades.test.ts

```diff
@@ -0,0 +1,134 @@
+import { expect } from 'chai'
+import { Mappings } from './wasm'
+import { ValueKind } from './wasm/runtime/ethereum'
+
+describe('onTrade', function () {
+  let mappings: Mappings
+
+  const user = `0x${'1337'.repeat(10)}`
+  const orderId = 1337n
+  const orderUid = `${user}-${orderId}`
+  const txHash = `0x${'01'.repeat(32)}`
+  const logIndex = 1n
+  const timestamp = 10n * 300n + 42n
+  const tradeId = `${txHash}-${logIndex}`
+
+  before(async () => {
+    mappings = await Mappings.load()
+
+    mappings.setEntity('Token', '0', {
+      id: '0',
+      address: `0x${'00'.repeat(20)}`,
+      fromBatchId: 0n,
+      symbol: 'OWL',
+      decimals: 18n,
+      name: 'Token OWL',
+      createEpoch: 0n,
+      txHash: `0x${'00'.repeat(32)}`,
+    })
+    mappings.setEntity('Order', `${user}-${orderId}`, {
+      id: orderUid,
+      owner: user,
+      orderId,
+      fromBatchId: 0n,
+      fromEpoch: 0n,
+      untilBatchId: 42n,
+      untilEpoch: 42n * 300n,
+      buyToken: '1',
+      sellToken: '0',
+      priceNumerator: 200000n * 10n ** 18n,
+      priceDenominator: 1000n * 10n ** 18n,
+      maxSellAmount: 200000n * 10n ** 18n,
+      minReceiveAmount: 1000n * 10n ** 18n,
+      soldVolume: 0n,
+      boughtVolume: 0n,
+      createEpoch: 0n,
+      cancelEpoch: null,
+      deleteEpoch: null,
+      txHash: `0x${'00'.repeat(32)}`,
+      txLogIndex: 0n,
+    })
+    mappings.setCallHandler((call) => {
+      const callName = `${call.contractName}.${call.functionName}`
+      switch (callName) {
+        case 'BatchExchange.currentPrices':
+          switch (call.functionParams[0].data) {
+            case 0n:
+              return [{ kind: ValueKind.Uint, data: 10n ** 18n }]
+            case 1n:
+              return [{ kind: ValueKind.Uint, data: 200n * 10n ** 18n }]
+          }
+          break
+        case 'BatchExchange.latestSolution':
+          return [
+            { kind: ValueKind.Uint, data: 9n }, // batch ID
+            { kind: ValueKind.Address, data: new Uint8Array(20) }, // submitter
+            { kind: ValueKind.Uint, data: 1337n }, // burnt fees
+            { kind: ValueKind.Uint, data: 42000n }, // objective value
+          ]
+      }
+
+      throw new Error(`unpexcted contract call ${JSON.stringify(call)}`)
+    })
+
+    mappings.onTrade(
```

Comment:
This is an interesting way of unit testing. You basically have one interaction in the before block and then only the asserts in each `it` block.

I'm usually not a fan of before blocks as they tend to setup global state that makes test cased dependent on one another. It works here since test cases don't actually do anything anymore, but curios to here your thoughts why you chose this instead of the good old. Setup act assert pattern?

Would we never want to test a different interactions?

Reply:
> but curios to here your thoughts why you chose this instead of the good old. Setup act assert pattern?

Mostly because of the duplicated code, it would be 80 lines of copy pasted code 3x. Basically this is a standard "add trade" process and just want to do a bunch of assertions on that chain of events.

> Would we never want to test a different interactions?

Yes, but then I expect them to go in different `describe` blocks and setup their own state for assertions. I imagine that at that point we could factor out the bits that can be reused into functions and just have different setups for the cases that need different interactions.